src/Middleware/ResponseCompression/src/ResponseCompressionBody.cs

```diff
@@ -59,6 +59,12 @@ internal async Task FinishCompressionAsync()
             {
                 await _compressionStream.DisposeAsync();
             }
+
+            // Adds the compression headers for HEAD requests even if the body was not used.
+            if (HttpMethods.IsHead(_context.Request.Method) && !_compressionChecked)
```

Comment:
Since most responses will have a body and set _compressionChecked, it's better to check that condition first.
```suggestion
            if (!_compressionChecked && HttpMethods.IsHead(_context.Request.Method))
```